clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -509,7 +542,9 @@ public boolean isChannelReady(String id) {
     private KafkaChannel channelOrFail(String id) {
         KafkaChannel channel = this.channels.get(id);
         if (channel == null)
-            throw new IllegalStateException("Attempt to retrieve channel for which there is no open connection. Connection id " + id + " existing connections " + channels.keySet());
+            channel = this.closedChannels.get(id);
+        if (channel == null)
```

Comment:
Is it safe to do this given that this method is called from `send`? I think I'd prefer to name this in a way that was obvious that it can return open or closed channels (unlike `channel`).